pkg/kn/flags/bool.go

```diff
@@ -0,0 +1,79 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package flags
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/pflag"
+)
+
+var negPrefix = "no-"
+
+// AddBothBoolFlags adds the given flag in both `--foo` and `--no-foo` variants.
+// If you do this, make sure you call ReconcileBoolFlags later to catch errors and
+// set the relationship between the flag values.
+func AddBothBoolFlags(f *pflag.FlagSet, p *bool, name, short string, value bool, usage string) {
+
+	negativeName := negPrefix + name
+
+	f.BoolVarP(p, name, short, value, usage)
+	f.Bool(negativeName, !value, "do not "+usage)
+
+	if value {
+		err := f.MarkHidden(name)
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		err := f.MarkHidden(negativeName)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+// ReconcileBoolFlags sets the value of the all the "--foo" flags based on
+// "--no-foo" if provided, and returns an error if both were provided.
+func ReconcileBoolFlags(f *pflag.FlagSet) error {
+	var err error
+	f.Visit(func(flag *pflag.Flag) {
+		// Return early from our comprehension
+		if err != nil {
+			return
+		}
+		// Walk the "no-" versions of the flags. Make sure we didn't set
+		// both, and set the positive value to the opposite of the "no-"
+		// value if it exists.
+		if strings.HasPrefix(flag.Name, "no-") {
+			positiveName := flag.Name[len(negPrefix):len(flag.Name)]
+			positive := f.Lookup(positiveName)
+			if positive.Changed {
+				err = fmt.Errorf("only one of --%s and --%s may be specified",
+					flag.Name, positiveName)
+				return
+			}
+			var noValue bool
+			noValue, err = strconv.ParseBool(flag.Value.String())
```

Comment:
I opened https://github.com/spf13/pflag/issues/214 upstream; I read the code and determined that this case is difficult-to-impossible to disable without deep surgery into the pflags code.

Reply:
So if I understand correctly, then we can't prevent the user from providing a value, so we need to check it ?

What if we just ignore the value ? (just asking, I see that `--no-foo=false` would be still possible, but if we respect this right now, it's hard to remove later when this (hopefully) becomes an error)

Reply:
I mean, couldn't we just check for `flag.Changed()` to check whether the option was provided instead of evaluating the value (which could be not want we want it to be) ?

Reply:
I see what you're saying, though I think you mean `.Changed`, and I think I disagree. While you _can_ provide `true` or `false`, having it not matter what you provide is even more confusing, I think.

Reply:
Or I guess I could guard against unexpected values...

Reply:
Yeah, guarding would be the best solution IMO (and btw, 'Changed()' probably doesn't fire when you use the default value. not sure, but had the impression when looking at the implementation). 

And then we should add tests for this behaviour, too.

Reply:
Tested it, and `Changed` works.